handlers/secrets.go

```diff
@@ -75,3 +86,238 @@ func makeSecretsArray(c *client.Client, secretNames []string) ([]*swarm.SecretRe
+				ownerLabel: ownerLabelValue,
+			},
+		},
+		Data: []byte(secret.Value),
+	})
+	if createSecretErr != nil {
+		return http.StatusInternalServerError, fmt.Errorf(
+			"error creating secret in secretPostHandler: %s",
+			createSecretErr,
+		)
+	}
+
+	return http.StatusCreated, nil
+}
+
+func updateSecret(c *client.Client, body []byte) (responseStatus int, err error) {
+	var secret requests.Secret
+
+	unmarshalErr := json.Unmarshal(body, &secret)
+	if unmarshalErr != nil {
+		return http.StatusBadRequest, fmt.Errorf(
+			"error unmarshaling secret in secretPutHandler: %s",
+			unmarshalErr,
+		)
+	}
+
+	foundSecret, getSecretErr, status := getSecretWithName(c, secret.Name)
+	if getSecretErr != nil {
+		return status, fmt.Errorf(
+			"cannot get secret with name: %s, which you want to remove. Error: %s",
+			secret.Name,
+			getSecretErr.Error(),
+		)
+	}
+
+	updateSecretErr := c.SecretUpdate(context.Background(), foundSecret.ID, foundSecret.Version, swarm.SecretSpec{
+		Annotations: swarm.Annotations{
+			Name: secret.Name,
+			Labels: map[string]string{
+				ownerLabel: ownerLabelValue,
+			},
+		},
+		Data: []byte(secret.Value),
+	})
+
+	if updateSecretErr != nil {
+		fmt.Println("updateSecretErr\n", updateSecretErr.Error())
+
+		return http.StatusInternalServerError, fmt.Errorf(
+			"couldn't update secret (name: %s, ID: %s) because of error: %s",
+			secret.Name,
+			foundSecret.ID,
+			getSecretErr.Error(),
+		)
+	}
+
+	return http.StatusOK, nil
+}
+
+func deleteSecret(c *client.Client, body []byte) (responseStatus int, err error) {
+	var secret requests.Secret
+
+	unmarshalErr := json.Unmarshal(body, &secret)
+	if unmarshalErr != nil {
+		return http.StatusBadRequest, fmt.Errorf(
+			"error unmarshaling secret in secretDeleteHandler: %s",
+			unmarshalErr,
+		)
+	}
+
+	foundSecret, getSecretErr, status := getSecretWithName(c, secret.Name)
+	if getSecretErr != nil {
+		return status, fmt.Errorf(
+			"cannot get secret with name: %s, which you want to remove. Error: %s",
+			secret.Name,
+			getSecretErr,
+		)
+	}
+
+	removeSecretErr := c.SecretRemove(context.Background(), foundSecret.ID)
+	if removeSecretErr != nil {
+		return http.StatusInternalServerError, fmt.Errorf(
+			"error trying to remove secret (name: `%s`, ID: `%s`): %s",
+			secret.Name,
+			foundSecret.ID,
+			removeSecretErr,
+		)
+	}
+
+	return http.StatusOK, nil
+}
+
+func MakeSecretsHandler(c *client.Client) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		body, readBodyErr := ioutil.ReadAll(r.Body)
+		if readBodyErr != nil {
+			log.Printf("couldn't read body of a request: %s", readBodyErr)
+
+			w.WriteHeader(http.StatusInternalServerError)
+
+			return
+		}
+
+		if r.Body != nil {
+			defer r.Body.Close()
+		}
+
+		var (
+			responseStatus int
+			responseBody   []byte
+			responseErr    error
+		)
+
+		switch r.Method {
+		case http.MethodGet:
+			responseStatus, responseBody, responseErr = getSecrets(c, body)
```

Comment:
Why is this method so different from the others? Could we keep it consistent for each handler like Lucas / Stefan have in their version?

https://github.com/openfaas/faas-netes/pull/344/files#diff-1d147804cec0e8b9c624b312ff9df6a4R37

Reply:
Done